addons/test_mail_full/tests/test_controller_attachment.py

```diff
@@ -35,3 +35,307 @@ def test_attachment_upload_portal(self):
                 (self.user_admin, True, sign),
             ),
         )
+
+    def test_independent_attachment_delete_portal(self):
+        """Test access to delete an attachment on portal"""
+        # Subtest format: (user, token, result, sign_token_params)
+        record = self.env["mail.test.portal"].create({"name": "Test"})
+        token, bad_token, sign, bad_sign, _ = self._get_sign_token_params(record)
+        self._execute_subtests_delete(
+            (
+                (self.guest, False, False, token),
+                (self.guest, True, False, token),
+                (self.guest, False, False, sign),
+                (self.guest, True, False, sign),
+                (self.guest, False, False, bad_token),
+                (self.guest, True, False, bad_token),
+                (self.guest, False, False, bad_sign),
+                (self.guest, True, False, bad_sign),
+                (self.user_admin, False, True, token),
+                (self.user_admin, True, True, token),
+                (self.user_admin, False, True, sign),
+                (self.user_admin, True, True, sign),
+                (self.user_admin, False, True, bad_token),
+                (self.user_admin, True, True, bad_token),
+                (self.user_admin, False, True, bad_sign),
+                (self.user_admin, True, True, bad_sign),
+                (self.user_employee, False, False, token),
+                (self.user_employee, True, False, token),
+                (self.user_employee, False, False, sign),
+                (self.user_employee, True, False, sign),
+                (self.user_employee, False, False, bad_token),
+                (self.user_employee, True, False, bad_token),
+                (self.user_employee, False, False, bad_sign),
+                (self.user_employee, True, False, bad_sign),
+                (self.user_portal, False, False, token),
+                (self.user_portal, True, False, token),
+                (self.user_portal, False, False, sign),
+                (self.user_portal, True, False, sign),
+                (self.user_portal, False, False, bad_token),
+                (self.user_portal, True, False, bad_token),
+                (self.user_portal, False, False, bad_sign),
+                (self.user_portal, True, False, bad_sign),
+                (self.user_public, False, False, token),
+                (self.user_public, True, False, token),
+                (self.user_public, False, False, sign),
+                (self.user_public, True, False, sign),
+                (self.user_public, False, False, bad_token),
+                (self.user_public, True, False, bad_token),
+                (self.user_public, False, False, bad_sign),
+                (self.user_public, True, False, bad_sign),
```

Comment:
There are definitely too many cases, let's see how much we can simplify this with itertools (see other comment) and then we'll decide if we need to reduce coverage a little bit.

Reply:
Don't do all combinations... if you code correctly checks tokens and signature, no need to check it for every kind of potential user. That's just adding noise, complexity and time for nothing. Try to cover specific areas that need coverage, don't add tests just because it is fun to do combinatorial explosions ^^ .